### PR TITLE
feat(iot): add Action to set a CloudWatch alarm

### LINKED ISSUE #17705: (iot): TopicRule action for CloudWatch alarms in AWS IoT
### Description

The CDK has no support to put CloudWatch alarm status via AWS IoT Rule.

### Use Case

Adding the action that puts CloudWatch alarm status to AWS IoT topic rule.

### Proposed Solution

Implementing action class.

Probably, following PRs help coding.

- https://github.com/aws/aws-cdk/pull/17503
- https://github.com/aws/aws-cdk/pull/17466
- https://github.com/aws/aws-cdk/pull/17307

### Other information

Follofing documentation will be help.

- https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iot-topicrule-cloudwatchalarmaction.html

It will be good first issue for learning implement feature in CDK!

### Acknowledge

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by jerry-shao**
Adding IoT Rule action for CloudWatch alarm.

Fixes #17705

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * The value of the alarm state.
+   */
+  readonly stateValue: string;
+}
+
+/**
+ * The action to change a CloudWatch alarm state.
+ */
+export class CloudWatchAlarmAction implements iot.IAction {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts:24]: Not sure I love this name. It's missing a verb. How about `CloudWatchSetAlarmAction`?

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Configuration properties of an action for CloudWatch alarm.
+ */
+export interface CloudWatchAlarmActionProps extends CommonActionProps {
+  /**
+   * The reason for the alarm change.
+   */
+  readonly stateReason: string;
+  /**
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts:15]: Missing empty line before this line.

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import { singletonActionRole } from './private/role';
+
+/**
+ * Configuration properties of an action for CloudWatch alarm.
+ */
+export interface CloudWatchAlarmActionProps extends CommonActionProps {
+  /**
+   * The reason for the alarm change.
+   */
+  readonly stateReason: string;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts:14]: 1. Let's just call this `reason`.
2. Seems like we can probably default this to something sensible...? I don't think we need to make this required.

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * The value of the alarm state.
+   */
+  readonly stateValue: string;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts:18]: 1. This should not be string. It should be of the type [`AlarmState` from the CloudWatch library](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cloudwatch.AlarmState.html).
2. I'm not in love with this name. How about `alarmStateToSet` instead?

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
+import { Template, Match } from '@aws-cdk/assertions';
+import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
+import * as iam from '@aws-cdk/aws-iam';
+import * as iot from '@aws-cdk/aws-iot';
+import * as cdk from '@aws-cdk/core';
+import { CloudWatchAlarmAction, CloudWatchAlarmActionProps } from '../../lib/cloudwatch-alarm-action';
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:6]: Imports in tests should use qualified imports to the library root, not to individual files:

```suggestion
import * as actions from '../../lib';
```

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  const stack = new cdk.Stack();
+  const topicRule = new iot.TopicRule(stack, 'MyTopicRule', {
+    sql: iot.IotSql.fromStringAsVer20160323("SELECT topic(2) as device_id, stateReason, stateValue FROM 'device/+/data'"),
+  });
+  const alarmArn = 'arn:aws:cloudwatch:us-east-1:123456789012:alarm:MyAlarm';
+  const alarm = cloudwatch.Alarm.fromAlarmArn(stack, 'MyAlarm', alarmArn);
+  const cloudWatchAlarmActionProps: CloudWatchAlarmActionProps = {
+    stateReason: '${stateReason}',
+    stateValue: '${stateValue}',
+  };
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:19]: Let's inline this variable.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+  // Then
+  Template.fromStack(stack).hasResourceProperties('AWS::IoT::TopicRule', {
+    TopicRulePayload: {
+      Actions: [
+        Match.objectLike({ CloudwatchAlarm: { StateReason: 'Test SetAlarmState' } }),
+      ],
+    },
+  });
+});
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:97]: I don't see the point of this test - it doesn't cover anything the previous one didn't.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+  // Then
+  Template.fromStack(stack).hasResourceProperties('AWS::IoT::TopicRule', {
+    TopicRulePayload: {
+      Actions: [
+        Match.objectLike({ CloudwatchAlarm: { StateValue: 'ALARM' } }),
+      ],
+    },
+  });
+});
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:123]: Same here (redundant test).

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+test('can set role', () => {
+  // Given
+  const stack = new cdk.Stack();
+  const topicRule = new iot.TopicRule(stack, 'MyTopicRule', {
+    sql: iot.IotSql.fromStringAsVer20160323("SELECT topic(2) as device_id, stateReason, stateValue FROM 'device/+/data'"),
+  });
+  const alarmArn = 'arn:aws:cloudwatch:us-east-1:123456789012:alarm:MyAlarm';
+  const alarm = cloudwatch.Alarm.fromAlarmArn(stack, 'MyAlarm', alarmArn);
+  const role = iam.Role.fromRoleArn(stack, 'MyRole', 'arn:aws:iam::123456789012:role/ForTest');
+  const cloudWatchAlarmActionProps: CloudWatchAlarmActionProps = {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:134]: Inline this variable too.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Awesome work @jerry-shao, thanks for the contribution! A few small comments, but in general looks amazing!

**Comment by skinny85** [comment]
BTW, you might want to update with the most recent `master`- that should make the build failure go away.

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts (hunk 1 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts:24]: Probably we can use the official CloudWatch API name: `SetAlarmState`.

So the class name could be `CloudWatchSetAlarmStateAction`

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts (hunk 2 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts:15]: Fixed

File: packages/@aws-cdk/aws-iot-actions/README.md [hunk 10]
```diff
@@ -149,6 +150,40 @@ const topicRule = new iot.TopicRule(this, 'TopicRule', {
...
+});
+
+const topicRule = new iot.TopicRule(this, 'TopicRule', {
+  sql: iot.IotSql.fromStringAsVer20160323("SELECT topic(2) as device_id FROM 'device/+/data'"),
+  actions: [
+    new actions.CloudWatchAlarmAction(alarm, {
+      stateReason: 'AWS Iot Rule action is triggered',
+      stateValue: 'ALARM',
+    }),
+  ]
```
[the remark this answers is not in the document]
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/README.md:183]: Fixed

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts (hunk 5 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:6]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts (hunk 6 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:19]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts (hunk 7 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:97]: Removed

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts (hunk 8 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:123]: Removed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts (hunk 9 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-alarm-action.test.ts:134]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts (hunk 4 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts:18]: Sounds good, using `cloudwatch.AlarmState` instead.

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts:24]: Sure. It's a little long, that's why I went with the shorter `CloudWatchSetAlarmAction`, but I don't hate `CloudWatchSetAlarmStateAction`.

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts (hunk 3 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-alarm-action.ts]: Fixed.

**Comment by skinny85** [comment]
@jerry-shao is this ready for another round of reviews?

**Comment by jerry-shao** [comment]
> @jerry-shao is this ready for another round of reviews?

Yup

**Comment by skinny85** [comment]
> > @jerry-shao is this ready for another round of reviews?
> 
> Yup

Ok. In the future, pleas re-request my review when you're done (there's a button in the top-right corner of the PR page, next to my avatar).

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-set-alarm-state-action.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      actions: ['cloudwatch:SetAlarmState'],
+      resources: [this.alarm.alarmArn],
+    }));
+
+    return {
+      configuration: {
+        cloudwatchAlarm: {
+          alarmName: this.alarm.alarmName,
+          roleArn: role.roleArn,
+          stateReason: this.props.reason ?? 'AWS IoT Rule action triggers a change',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-set-alarm-state-action.ts:46]: I was thinking of a default more like `` `Set state of '${this.alarm.alarmName}' to '${props.alarmStateToSet}'` `` 🙂.

File: packages/@aws-cdk/aws-iot-actions/README.md [hunk 12]
```diff
@@ -149,6 +150,39 @@ const topicRule = new iot.TopicRule(this, 'TopicRule', {
...
+import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
+import * as iot from '@aws-cdk/aws-iot';
+import * as actions from '@aws-cdk/aws-iot-actions';
+
+const metric = new cloudwatch.Metric({
+  namespace: 'MyNamespace',
+  metricName: 'MyMetric',
+  dimensions: { MyDimension: 'MyDimensionValue' },
+});
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/README.md:167]: We can probably kill this empty line.

File: packages/@aws-cdk/aws-iot-actions/README.md [hunk 13]
```diff
@@ -149,6 +150,39 @@ const topicRule = new iot.TopicRule(this, 'TopicRule', {
 });
 ```
 
+## Change the state of an Amazon CloudWatch alarm
+
+The code snippet below creates an AWS IoT Rule that changes the state of an Amazon CloudWatch alarm when it is triggered.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/README.md:155]:
```suggestion
The code snippet below creates an AWS IoT Rule that changes the state of an Amazon CloudWatch alarm when it is triggered:
```

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+  // When
+  topicRule.addAction(
+    new actions.CloudWatchSetAlarmStateAction(
+      alarm,
+      {
+        reason: 'Test reason',
+        alarmStateToSet: cloudwatch.AlarmState.ALARM,
+      },
+    ),
+  );
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts:25]:
```suggestion
  topicRule.addAction(new actions.CloudWatchSetAlarmStateAction(alarm, {
    reason: 'Test reason',
    alarmStateToSet: cloudwatch.AlarmState.ALARM,
  }));
```

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+  topicRule.addAction(
+    new actions.CloudWatchSetAlarmStateAction(
+      cloudwatch.Alarm.fromAlarmArn(stack, 'MyAlarm', 'arn:aws:cloudwatch:us-east-1:123456789012:alarm:MyAlarm'),
+      {
+        reason: '${stateReason}',
+        alarmStateToSet: cloudwatch.AlarmState.ALARM,
+        role: iam.Role.fromRoleArn(stack, 'MyRole', 'arn:aws:iam::123456789012:role/ForTest'),
+      },
+    ),
+  );
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts:93]:
```suggestion
  topicRule.addAction(new actions.CloudWatchSetAlarmStateAction(
    cloudwatch.Alarm.fromAlarmArn(stack, 'MyAlarm', 'arn:aws:cloudwatch:us-east-1:123456789012:alarm:MyAlarm'),
    {
      reason: '${stateReason}',
      alarmStateToSet: cloudwatch.AlarmState.ALARM,
      role: iam.Role.fromRoleArn(stack, 'MyRole', 'arn:aws:iam::123456789012:role/ForTest'),
    },
  ));
```

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+  // When
+  topicRule.addAction(
+    new actions.CloudWatchSetAlarmStateAction(
+      alarm,
+      {
+        alarmStateToSet: cloudwatch.AlarmState.ALARM,
+      },
+    ),
+  );
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts:126]:
```suggestion
  topicRule.addAction(new actions.CloudWatchSetAlarmStateAction(alarm, {
    alarmStateToSet: cloudwatch.AlarmState.ALARM,
  }));
```

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+
+    const alarm = new cloudwatch.Alarm(this, 'MyAlarm', {
+      metric: metric,
+      threshold: 100,
+      evaluationPeriods: 3,
+      datapointsToAlarm: 2,
+    });
+
+    topicRule.addAction(new actions.CloudWatchSetAlarmStateAction(alarm, {
+      reason: 'Test reason',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts:30]: Let's test out the default here, now that I suggested it to be more complicated 🙂.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+const app = new cdk.App();
+
+class TestStack extends cdk.Stack {
+  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
+    super(scope, id, props);
+
+    const topicRule = new iot.TopicRule(this, 'TopicRule', {
+      sql: iot.IotSql.fromStringAsVer20160323("SELECT topic(2) as device_id FROM 'device/+/data'"),
+    });
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts:14]: Let's create the `TopicRule` after the Alarm?

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+
+    const metric = new cloudwatch.Metric({
+      namespace: 'MyNamespace',
+      metricName: 'MyMetric',
+      dimensions: { MyDimension: 'MyDimensionValue' },
+    });
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts:21]: No need for this empty line I think either.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks fantastic @jerry-shao! A few last tiny comments, but this is in a great place already.

File: packages/@aws-cdk/aws-iot-actions/README.md (hunk 12 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/README.md:167]: Fixed

File: packages/@aws-cdk/aws-iot-actions/README.md (hunk 13 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/README.md:155]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts (hunk 14 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts:25]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts (hunk 16 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts:126]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts (hunk 19 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts:21]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts (hunk 15 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/cloudwatch-set-alarm-state-action.test.ts:93]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts (hunk 18 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts:14]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-set-alarm-state-action.ts (hunk 11 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/lib/cloudwatch-set-alarm-state-action.ts:46]: Fixed.

File: packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts (hunk 17 above)
Comment (jerry-shao) [reply packages/@aws-cdk/aws-iot-actions/test/cloudwatch/integ.cloudwatch-set-alarm-state-action.ts:30]: Fixed.

**Review by skinny85 — APPROVED** [review]
Looks fantastic, thanks so much for the contribution @jerry-shao!

**Comment by jerry-shao** [comment]
> Looks fantastic, thanks so much for the contribution @jerry-shao!

Thanks @skinny85 , any chance you could trigger the workflows?

**Comment by skinny85** [comment]
> > Looks fantastic, thanks so much for the contribution @jerry-shao!
> 
> Thanks @skinny85 , any chance you could trigger the workflows?

Done!

**Pull request merged**